keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1717,45 +1618,76 @@ type ResourceClaimTemplateSpec struct {
-// locate the parameters for a ResourceClass.
-type ResourceClassParametersReference struct {
-    // APIGroup is the group for the resource being referenced. It is
-    // empty for the core API. This matches the group in the APIVersion
-    // that is used when creating the resources.
-    // +optional
-    APIGroup string
-    // Kind is the type of resource being referenced. This is the same
-    // value as in the parameter object's metadata.
-    Kind string
-    // Name is the name of resource being referenced.
-    Name string
-    // Namespace that contains the referenced resource. Must be empty
-    // for cluster-scoped resources and non-empty for namespaced
-    // resources.
-    // +optional
-    Namespace string
+// Quota controls whether a ResourceClaim may get allocated.
+// Quota is namespaced and applies to claims within the same namespace.
+type Quota struct {
+	metav1.TypeMeta
+	// Standard object metadata.
+	metav1.ObjectMeta
+
+	Spec QuotaSpec
 }
 
-// ResourceClaimParametersReference contains enough information to let you
-// locate the parameters for a ResourceClaim. The object must be in the same
-// namespace as the ResourceClaim.
-type ResourceClaimParametersReference struct {
-    // APIGroup is the group for the resource being referenced. It is
-    // empty for the core API. This matches the group in the APIVersion
-    // that is used when creating the resources.
-    // +optional
-    APIGroup string
-    // Kind is the type of resource being referenced. This is the same
-    // value as in the parameter object's metadata, for example "ConfigMap".
-    Kind string
-    // Name is the name of resource being referenced.
-    Name string
+type QuotaSpec struct {
+	// AllowManagementAccess controls whether claims with ManagementAccess
+	// may be allocated. If multiple quota objects exist and at least one
+	// has a true value, access will be allowed. The default to deny such access.
+	//
+	// +optional
+	AllowManagementAccess bool `json:"allowManagementAccess,omitempty"`
+
+	// Stretch goals for 1.31:
+	//
+	// - maximum number of devices matching a selector
+	// - maximum sum of a certain quantity attribute
+	//
+	// These are additional requirements when checking whether a device
+	// instance can satisfy a request. Creating a claim is always allowed,
+	// but allocating it fails when the quota is currently exceeded.
+
+	// Other useful future extensions (>= 1.32):
+
+	// DeviceLimits is a CEL expression which take the currently allocated
+	// devices and their attributes and some new allocations as input and
+	// returns false if those allocations together are not permitted in the
+	// namespace.
+	//
+	// DeviceLimits string
+
+	// A class listed in DeviceClassDenyList must not be used in this
+	// namespace. This can be useful for classes which contain
+	// configuration pieces that a user in this namespace should not have
+	// access to.
+	//
+	// DeviceClassDenyList []string
```

Comment:
covered by "device count in class"...except with the "all" Amount, we are back to allocation time vs admission time. We may have to live with that in 1.31 (admission time only) - admin can use an admission webhook to not allow "all" maybe as a workaround.